[Qt] Add transaction rate trend graph in debug dialog

Justaphf opened the pull request:
This is a customization I added to my client that I figured I'd put a PR in for.  This is based on the existing network traffic graph and basically expands upon the existing smoothed 60-second average and runtime peak transaction rate statistics displayed on the first tab of the debug dialog in the Qt client.  The trend graph is based on taking 1-second samples of the smoothed and instantaneous rates and storing up to 24-hours worth of samples.  Samples older than 24-hours are purged from memory.  In addition to the graph of the instantaneous and smoothed samples, some high level statistics are displayed giving the runtime, sample size (24-hour), and display window (dynamic) instantaneous peak values and smoothed average values.

Unlike the network traffic graph, the transaction rate trend graph maintains up to 24 hours of samples in memory allowing changing the display time window without losing the previous samples.  The displayed window can be as short as the last 5 minutes or as long as the full 24 hours of samples.

The main changes related to this PR are as follows:
1. Expose some of the method local constants for use external to `txmempool.UpdateTransactionsPerSec()` related to smoothing window size (60 seconds) and resolution of the instantaneous rate (1 second).  These are used to set some of the display strings as well as control the sampling timer to ensure we don't attempt to gather samples at a higher frequency than the instantaneous resolution.

2. Expose most recently computed instantaneous transaction rate computed in `txmempool.UpdateTransactionsPerSec()` to use as a sample point in the trend graph.  This allows displaying the instantaneous rate which will show when there are gaps in transactions (when zoomed in to a small enough window).  The instantaneous rate is overlaid with the smoothed average on the graph.

3. Refactor `txmempool.UpdateTransactionsPerSec()` to allow updating the smoothed average without adding a new transaction.  This fixes an issue where if there are gaps between transactions the smoothed rate would not decay and would falsely appear to be at a plateau until the next transaction was added to the commit queue.  This was very obvious when looking at the samples on the graph if there were several seconds between transactions.

4. Refactor accessor methods for the individual statistics to instead return all 3 stats (smoothed average, instantaneous value, peak value) to instead all be returned atomically as a group.  This reduces the taking/releasing of the `cs_txPerSec` lock when displaying the statistics.  NOTE: The `cs_txPerSec` critical section is using `std::lock_guard` instead of the `LOCK` macro, and as such I made the helper method require the lock be passed in as there is no equivalent to `AssertLockHeld` and this avoids the need to change the lock type (though perhaps it should be changed).

5. Add the new transaction rate widget to the debug dialog.  When at larger sample window sizes, all sample points which end up mapping to the same pixel x value are aggregated.  For the instantaneous rate, the max sample is used, for the smoothed average, the average of all samples is used.

Some sample images from recent runs.  I took these screen captures at minimum size of the dialog for inclusion here, but they look much better on a wide screen monitor fully stretched horizontally across the screen.

5-minute window:
<img width="451" alt="TxnRate_5m" src="https://user-images.githubusercontent.com/6402604/65295585-d3827980-db2f-11e9-947a-9572348a518d.png">

1-hour window:
<img width="451" alt="TxnRate_1h" src="https://user-images.githubusercontent.com/6402604/65295604-de3d0e80-db2f-11e9-9e85-b2d78afe7a4c.png">

24-hour window:
<img width="451" alt="TxnRate_24h" src="https://user-images.githubusercontent.com/6402604/65295623-ea28d080-db2f-11e9-91c4-1e8dabf2d1a7.png">

Based on having used this for a couple of weeks now, I have discovered some areas that I believe could be improved upon in subsequent versions.
1. The sample window is locked to always show the most recent samples.  To see older samples requires increasing the window size which loses the finer detail.  It would be useful to have a horizontal scroll bar to allow seeing older samples at a smaller window size.

2. The sample window slider is 5 minutes per tick all the way across.  I feel that this is probably only useful for shorter windows (1 hour or less).  Make the ticks step size dynamic base on window size.

3. When there are large spikes, the y-axis is scaled to show the peak of this spike.  This can lose detail of most of the data and it would be useful to be able to set a maximum scale on the y-axis, truncating the tops of spikes.  Perhaps an option to set the maximum y-axis value based either on the peak or the 95th percentile value.

4. I don't really see any issues with resource usage with this, but maybe it would make sense to have a command line/config argument to turn this feature off

5. Related to 4 above, I have not tested this during high transaction throughput.  This could potentially run into issues in that case due to the locking of `cs_txPerSec`.  I'm interested to see what the trend graph looks like during the next GTI run.

Justaphf committed:
Add transaction rate graph to Qt debug UI

This adds a graph of (up to) the last 24-hours of transaction rate data
in the Qt debug GUI.

The instantaneous (1-second resolution) transaction rate shows up in
yellow on the graph.  This is useful for seeing lulls in transaction
activity.

The smoothed (60-second window) transaction rate data shows up in green
on the graph.  This corresponds to the smoothed transaction rate
presently displayed on the main tab of the debug dialog.

Limits sample frequency to no more than the instantaneous resolution
by adding a new timer with the same frequency as the instantaneous
resolution.  This is to prevent the graph from showing flat steps for samples
taken within the same sample period, which occurred when I had this using
the 0.5 second update frequency of the existing timer.

As part of this change, use the consts for instantaneous resolution and
smoothing sample window size to:
1. Limit UI sample rate to not exceed instantaneous resolution via new
update timer specific to transaction rate.
2. Update the UI to set the sample resolution & smoothing window size in
the labels at startup
3. Update graph widget to use the new sample frequency instead of the
half second frequency.
4. Don't initiate sampling until the peak transaction rate is recorded as being
higher than 0.  This is to exclude the 10 second period during start-up
where the transaction rate statistics are ignored so as not to have a large
spike due to the loading of mempool.dat file.

Justaphf committed:
Refactor to return all txn rate stats atomically as a group

This change serves two purposes:
1. Group the return of all transaction rate statistics in a single
method in order to reduce the number of times the cs_txPerSec lock is
taken.  These statistics are commonly requested together and the old
method of individually returning each stat would cause frequent taking
and releasing of the lock in quick succession.  By grouping the return
under a single lock, this also ensures that the statistics are all
returned atomically.  As part of this, the individual stat return
methods have been removed to discourage misuse. This is necessary as the
stats will be retrieved at regular intervals now for display in the
transaction rate graph (to be added in subsequent commit).

2. Allow updating transaction rate stats without adding a new
transaction.  This is to allow continual decay of the smoothed
transaction rate and re-computation of the instantaneous rate during
periods of time longer than the instantaneous resolution when no
transactions are received.  Previously if no transactions were received
the smoothed transaction rate would incorrectly show as plateaued at the
rate computed when the last transaction was received.

This issue is much more obvious when viewed in a trend chart (subsequent
commit in this feature adds that trend chart to the Qt GUI).

Justaphf committed:
Expose some local variables in the update txn rate method

The mempool computes an instantaneous transaction rate at 1-second
resolution in order to update the peak transaction rate for the entire
runtime of the node.  This instantaneous rate was being thrown away
after updating the peak.  This commit exposes the most recent
instantaneous rate to be used for charting.
Additionally exposes the smoothing duration and instantaneous transaction
resolution constants for external use.

Justaphf committed:
gitignore Visual Studio .vs directory

ptschip commented:
Awesome, been wanting this for a while!  Running it now and will look over the code in the next few days.

ptschip commented:
Initial observation:  do we really need to see the "Displayed" data on the transaction graph? isn't Runtime and 24 hour enough?

Justaphf commented:
Thanks for looking it over!

I added the "Displayed" data to show the peak and average values for exactly and only the samples that are used to draw the on-screen graph.  Not sure how useful others will find it, but I wanted the legend to contain a "this is exactly what you are seeing on-screen now" metric in addition to the longer term trend values.  Perhaps a better/dynamic label name would be good?  With the present implementation the graph is locked to always show the most recent samples so it should be feasible to dynamically set the label to be "X-Hours" where X is the zoom level converted to hours?

Based on my actual usage over the past month or so, what happens a lot is we randomly get a very tall spike which dominates the values for the next 24-hours until it falls out of the sample set.  What I find myself doing a lot is looking at the full 24-hours worth of samples, then changing my zoom to something shorter than when the spike occurred to see what things were like immediately after the spike cleared (i.e. what's the next highest spike, what's the sustained average when the spike has decayed out, etc).  A lot of this has to do with my 3rd point above about losing a lot of the finer detail visually when a spike is on-screen.

ptschip commented on src/txmempool.h:
```diff
@@ -35,6 +35,10 @@ inline bool AllowFree(double dPriority)
+static const double TRANSACTION_RATE_SMOOTHING_WINDOW_SECONDS = 60;
+/** Sample resolution in milliseconds over which to compute the instantaneous transaction rate */
+static const int INSTANTANEOUS_TRANSACTION_RATE_RESOLUTION_MILLISECONDS = 1000;
 
```
nit: seems like a very long variable name, what about just TRANSACTION_RATE_RESOLUTION? . The comment above already says it's in millis, or maybe TX_RATE_RESOLUTION_MILLIS?  same with TRANSACTION_RATE_SMOOTHING, maybe just use TX instead of TRANSACTION.

Justaphf replied:
Going to shorten names to `TX_RATE_RESOLUTION_MILLIS` and `TX_RATE_SMOOTHING_SEC`.

ptschip commented:
What about making the default window 24hrs rather than just 1 hr.

ptschip reviewed (approved):
Been running for a while. I think this is great and have a nit comment but I don't want to hold up this PR just for that since you're probably going to be making some upgrades later anyway.

Justaphf commented:
@ptschip I'm going to quickly make the couple of changes you suggested, as well as rebasing on latest dev and then force pushing.  Hadn't really gotten time to work on it since I put the PR in due to external time constraints but I have a couple of hours tonight to clean this up.  Thanks for looking through it.

Justaphf commented:
Latest push includes:
1. Rebased on top of latest `dev` branch.
2. Refactored all variable/const names which had time units in the name to standardize on `millis` for milliseconds and `sec` for seconds.
3. Set the default window size to 24 hours instead of 1 hour.
4. Updated the "Displayed" labels to be dynamically set based on the specified view window.  Uses the form "X.XX-Hours" if >= 1 hour window size, or "XX-Minutes" if < 1 hour.  NOTE: This does cause a redundant double entry when at exactly 24-hours.  Maybe this should be changed to hide the values if displayed window value is the same as the sample window size.
5. Updated the window slider to be in steps of 5 minutes when < 1 hour, and steps of 15 minutes when > 1 hour.  This is both because I believe that 5 minute steps for a window size larger than 1 hour doesn't really make sense, and because 15 minute steps guarantee that minutes converted to hours will always need 2 or less decimals precision to accurately display value.

NOTE: As part of this change set I groomed down to the original 4 commits with the changes mentioned above fixed up into those commits.

Justaphf commented:
Ha, now that the variable names are shorter the linter is complaining about multi-lines needing to becoming single lines again.  Will fix and push again.

ptschip reviewed (approved):
Been using this all week, really love it!

gandrewstone commented on src/qt/transactiongraphwidget.cpp:
```diff
@@ -0,0 +1,376 @@
+{
+    int h = height() - YMARGIN * 2, w = width() - XMARGIN * 2;
+    int x = XMARGIN + w, yAvg, yPeak;
+    int samplesInWindow = getSamplesInDisplayWindow();
```
use max(1, getSamplesInDisplayWindow()) to ensure no div by zero on line 80

Justaphf replied:
#1960 
I corrected in a different manner than using std::max() though.

gandrewstone commented on src/qt/transactiongraphwidget.cpp:
```diff
@@ -0,0 +1,376 @@
+    // set
+    float tSumAll = 0.0f, tSumDisplay = 0.0f;
+    int countAll = vSmoothedSamples.size();
+    int countDisplay = countAll < numSamples ? countAll : numSamples;
```
make sure this is at least 1 to avoid div by 0 in line 316 and 348

Justaphf replied:
#1960 
Corrected by explicitly checking for > 0 in those two locations before attempting division.

gandrewstone commented:
I can't trigger the 2 possible div by 0 problems that I found, probably because various start up and screen size limits make them not possible in practice.  So I'm going to merge this excellent tool, but would like to see a small PR that addresses the possible div by 0s at the point where they may occur.  This will ensure that if subsequent changes (possibly by engineers who do not understand the full repercussions) change the behavior enough to allow 0s in these fields we won't get a crash.

ptschip commented:
Looking at the current display it looks like you have "24 hours" listed twice

<img width="130" alt="Capture" src="https://user-images.githubusercontent.com/12862928/66141789-23921f00-e5b9-11e9-97eb-a78777942b80.PNG">

ptschip commented:
never mind, i get it...it changes when you modify the timeline
